src/functor.jl

```diff
@@ -1,5 +1,13 @@
+function functor(T, x)
+  names = fieldnames(T)
+  if isempty(names)
+    return (), _ -> x
+  end
+  S = T.name.wrapper # remove parameters from parametric types
+  vals = ntuple(i -> getfield(x, names[i]), length(names))
+  return NamedTuple{names}(vals), y -> S(y...)
+end
```

Comment:
I think this will be slow. In https://github.com/FluxML/Flux.jl/pull/1932 it needed a generated function to be as quick as before:

https://github.com/FluxML/Flux.jl/pull/1932/files#diff-34ea0dd342eeb0864015f5ab93bd9db67f6f46ec46d4bf8dbebe9b8c1a433b04R129-R141

Reply:
let's keep this in mind for a future optimization-oriented PR

Reply:
@mcabbott I'm not familiar with generated function. Can you show me the snippet we should use here? Otherwise it will be done in a future PR.

Reply:
My link above is a few lines off. I don't remember all the details, but I think my comments are saying that it works around `Base.typename(T).wrapper` taking 2μs every time. However, this PR doesn't do that, it calls `S = constructorof(T)`. Maybe that removes the need for `@generated`. Might be worth timing these two?

Reply:
Trying to time this today...
```julia
julia> struct Mine1{T,S}
       x::T
       y::S
       end

julia> x = Mine1([1,2,3], [4,5,6]')
Mine1{Vector{Int64}, Adjoint{Int64, Vector{Int64}}}([1, 2, 3], adjoint([4, 5, 6]))

julia> @btime Base.typename(typeof($x)).wrapper
  1.458 ns (0 allocations: 0 bytes)
Mine1

julia> @btime ConstructionBase.constructorof(typeof($x))
  1.458 ns (0 allocations: 0 bytes)
Mine1

julia> VERSION
v"1.12.0-DEV.1375"
```
and on Julia 1.6 -- slower but far from 2μs, even if I have a faster computer than I did then...
```julia
julia> @btime Base.typename(typeof($x)).wrapper
  14.194 ns (0 allocations: 0 bytes)
Mine1

julia> @btime ConstructionBase.constructorof(typeof($x))
  0.001 ns (0 allocations: 0 bytes)
Mine1

julia> VERSION
v"1.6.0"
```